### PR TITLE
Fixed #1262

### LINKED ISSUE #1262: Spinning Bot Zombie Games
Reproduce:
- Join 1.9 lobby
- Join a bot
- Select a map you do not have
- Start the game
- When the prompt to download missing map, either cancel or hit okay on the prompt

Notice:
- You are disconnected from the game, but there is still a zombie player playing in the bot, and the game will just spin through rounds
- You can reconnect to the game, but there is no way to stop the bot unless you kill all of the client processes

### PR DESCRIPTION
**Pull request by RoiEXLab**
Adresses  #1262
The problem was actually a different one, but somehow related.
There were 2 kind of problems, the first being introduced (I think) by my try statement inside the event queue, because the exception was caucght by that and not by outer try clauses. The second one being a headless opened process, which had no frame and didn't close by itself.
Both are now fixed

### CODE REVIEW

File: src/games/strategy/engine/framework/GameRunner.java [hunk 1]
```diff
@@ -666,21 +676,34 @@ public static void joinGame(final GameDescription description, final Messengers
...
     final List<String> commands = new ArrayList<>();
-    ProcessRunnerUtil.populateBasicJavaArgs(commands, newClassPath);
-    commands.add("-D" + TRIPLEA_CLIENT_PROPERTY + "=true");
-    commands.add("-D" + TRIPLEA_PORT_PROPERTY + "=" + port);
-    commands.add("-D" + TRIPLEA_HOST_PROPERTY + "=" + hostAddressIP);
-    commands.add("-D" + TRIPLEA_NAME_PROPERTY + "=" + messengers.getMessenger().getLocalNode().getName());
-    final String javaClass = "games.strategy.engine.framework.GameRunner";
-    commands.add(javaClass);
-    ProcessRunnerUtil.exec(commands);
+    final boolean startInNewProcess = !("".equals(newClassPath) && MainFrame.getInstance() != null);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: ` !("".equals(newClassPath) && MainFrame.getInstance() != null);`
Can be written more simply as:
`newClassPath.isEmpty() || MainFrame.getInstance() == null`

You'll need to double check that `newClassPath` is not null, or add a null check to that above conditional.

File: src/games/strategy/engine/framework/GameRunner.java [hunk 2]
```diff
@@ -666,21 +676,34 @@ public static void joinGame(final GameDescription description, final Messengers
...
+    String prefix = startInNewProcess ? "-D" : "";
+    if (startInNewProcess) {
+      ProcessRunnerUtil.populateBasicJavaArgs(commands, newClassPath);
+    }
+    commands.add(prefix + TRIPLEA_CLIENT_PROPERTY + "=true");
+    commands.add(prefix + TRIPLEA_PORT_PROPERTY + "=" + port);
+    commands.add(prefix + TRIPLEA_HOST_PROPERTY + "=" + hostAddressIP);
+    commands.add(prefix + TRIPLEA_NAME_PROPERTY + "=" + messengers.getMessenger().getLocalNode().getName());
+    if (startInNewProcess) {
+      final String javaClass = "games.strategy.engine.framework.GameRunner";
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Please replace the hardcoded class name reference "games.strategy.engine.framework.GameRunner" with the equivalent of `GameRunner.class.getName()`

File: src/games/strategy/engine/framework/ProcessRunnerUtil.java [hunk 3]
```diff
@@ -101,13 +102,11 @@ public static void exec(final List<String> commands) {
...
-          while (s.read() >= 0) {
-            // just read
-          }
-        } catch (final IOException e) {
-          ClientLogger.logQuietly(e);
+        Scanner scanner = new Scanner(s);
+        while (scanner.hasNextLine()) {
+          System.out.println(scanner.nextLine());
         }
+        scanner.close();
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/ProcessRunnerUtil.java]: Can use a java7 autoclose block and remove the `scanner.close()`
```
try(Scanner scanner = new Scanner(s) {

}
```

File: src/games/strategy/engine/framework/ProcessRunnerUtil.java [hunk 4]
```diff
@@ -101,13 +102,11 @@ public static void exec(final List<String> commands) {
       // we need to read the input stream to prevent possible
       // deadlocks
       final Thread t = new Thread(() -> {
-        try {
-          while (s.read() >= 0) {
-            // just read
-          }
-        } catch (final IOException e) {
-          ClientLogger.logQuietly(e);
+        Scanner scanner = new Scanner(s);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/ProcessRunnerUtil.java]: Why send the output to standard out?
Are you sure there will be no blocking issues readying bytes vs whole lines?

File: src/games/strategy/triplea/ResourceLoader.java [hunk 5]
```diff
@@ -184,7 +184,7 @@ public URL getResource(final String inputPath) {
     String path = resourceLocationTracker.getMapPrefix() + inputPath;
     return getMatchingResources(path).stream().findFirst().orElse(
         getMatchingResources(inputPath).stream().findFirst().orElseGet(
-            ()->null));
+            () -> null));
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ResourceLoader.java]: Small annoyance - there are a number of unrelated changes in this PR. Particularly for bug fixes, but also in general, please put unrelated changes on their own branches.

File: src/games/strategy/triplea/TripleA.java [hunk 6]
```diff
@@ -90,7 +92,7 @@ public void shutDown() {
 
 
   @Override
-  public void startGame(final IGame game, final Set<IGamePlayer> players, final boolean headless) {
+  public void startGame(final IGame game, final Set<IGamePlayer> players, final boolean headless, Runnable exceptionAction) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/TripleA.java]: Instead of a 'Runnable exceptionAction', what if we instead returned 'true/false' to indicate if the game started? Would that simplify things?

It also seems the headless instance will need cleanup, a default action of no-op for cleanup does not seem right for all cases.

File: src/games/strategy/triplea/TripleA.java [hunk 7]
```diff
@@ -115,18 +117,25 @@ public void startGame(final IGame game, final Set<IGamePlayer> players, final bo
...
+          soundChannel = new DefaultSoundChannel(localPlayers);
+          game.addSoundChannel(soundChannel);
+          frame.setSize(700, 400);
+          frame.setVisible(true);
+          ClipPlayer.play(SoundPath.CLIP_GAME_START);
+          connectPlayers(players, frame);
+          frame.setExtendedState(Frame.MAXIMIZED_BOTH);
+          frame.toFront();
+        } catch (Exception e) {
+          if(game instanceof ClientGame){
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/TripleA.java]: `instanceof` checks are always a major design smell. There are generally better ways to accomplish the same, notably via polymorphism. But more to point, if a caller passed in an exception action, it probably wants it to be executed. In addition, we call ClientLogger in all types of code, it defaults to printing to System.out if there is no Swing UI to print to. It wouldn't hurt to verify this, but ClientLogger should be safe to call in just about  any context.

File: src/games/strategy/engine/xml/TestGameLoader.java [hunk 8]
```diff
@@ -35,7 +35,7 @@
    * The game is about to start.
    */
   @Override
-  public void startGame(final IGame game, final Set<IGamePlayer> players, final boolean headless) {}
+  public void startGame(final IGame game, final Set<IGamePlayer> players, final boolean headless, Runnable exceptionAction) {}
```
Comment (DanVanAtta) [inline src/games/strategy/engine/xml/TestGameLoader.java]: This interface could use some reworking, the boolean parameter for 'headless' is a hack.

Really we have these two APIs bundled into just one:
```
public void startGame(final IGame game, final Set<IGamePlayer> players);
public void startHeadlessGame(final IGame game, final Set<IGamePlayer> players)
```

Likely that split is a first step. Then I would aggregate the stuff in common, and restructure to use composition to execute the common parts.

For now though I think it would be better to return a status result, instead of adding a parameter that is sometimes used, sometimes not.

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java [hunk 9]
```diff
@@ -335,35 +335,38 @@ private void startGameInNewThread(final byte[] gameData, final Map<String, INode
...
-        SwingUtilities.invokeLater(() -> JOptionPane.getFrameForComponent(m_ui).setVisible(false));
-        try {
-          // game will be null if we loose the connection
-          if (m_game != null) {
-            try {
-              data.getGameLoader().startGame(m_game, playerSet, false);
-              data.testLocksOnRead();
-            } catch (final Exception e) {
-              ClientLogger.logQuietly(e);
+    new Thread(() -> {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: We call `startGameInNewThread` from within a new thread. Is creating a new thread here really necessary? Does it change the exception handling picture?

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java [hunk 10]
```diff
@@ -335,35 +335,38 @@ private void startGameInNewThread(final byte[] gameData, final Map<String, INode
...
+      SwingUtilities.invokeLater(() -> JOptionPane.getFrameForComponent(m_ui).setVisible(false));
+      try {
+        // game will be null if we loose the connection
+        if (m_game != null) {
+          try {
+            data.getGameLoader().startGame(m_game, playerSet, false, () -> {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: Brownie points if you replace `false` with an enum. Having looked at the method API just a moment ago, already I've forgotten if it was 'isHeadless', or 'isClient'. APIs that require you to look at their implementation/signature/javadocs to simply use them, are not usable, and scale poorly when it comes to maintenance cost. Namely whenever anyone reads this line, you have to jump around to somewhere else to understand it. Then repeat for each of the following lines.. who knows what exactly happens in game shutdown vs messenger shutdown and loading window done wait..

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java [hunk 11]
```diff
@@ -335,35 +335,38 @@ private void startGameInNewThread(final byte[] gameData, final Map<String, INode
...
               m_game.shutDown();
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: 1. It's a pretty blatant copy-paste to repeat these 4 lines. You could at least declare a `Runnable closeAction` instance, and call `closeAction.run()` in the exception handler to avoid the duplication.
2. If `startGame` swallows any exceptions thrown, and executes our exception action, why are we catching an exception here at all? It seems like the catch block here is dead code, which was the bug to begin with (that the shutdown code was no longer being executed, and will never be executed)

File: src/games/strategy/engine/framework/GameRunner.java [hunk 12]
```diff
@@ -633,7 +642,8 @@ public static void joinGame(final GameDescription description, final Messengers
     }
     final Version engineVersionOfGameToJoin = new Version(description.getEngineVersion());
     String newClassPath = null;
-    if (!ClientContext.engineVersion().getVersion().equals(engineVersionOfGameToJoin)) {
+    final boolean areSameVersion = ClientContext.engineVersion().getVersion().equals(engineVersionOfGameToJoin);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: nit, `areSameVersion` is a really awkward name. `sameVersion` is shorter, and would read much better IMO

File: src/games/strategy/engine/framework/GameRunner.java [hunk 13]
```diff
@@ -666,21 +676,34 @@ public static void joinGame(final GameDescription description, final Messengers
...
+    commands.add(prefix + TRIPLEA_NAME_PROPERTY + "=" + messengers.getMessenger().getLocalNode().getName());
+    if (startInNewProcess) {
+      final String javaClass = "games.strategy.engine.framework.GameRunner";
+      commands.add(javaClass);
+    }
+    if (startInNewProcess) {
+      ProcessRunnerUtil.exec(commands);
+    } else {
+      MainFrame.getInstance().setVisible(true);
+      handleCommandLineArgs(commands.toArray(new String[commands.size()]), COMMAND_LINE_ARGS, GameMode.SWING_CLIENT);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: The path to having this called is a bit twisted. It also seems like we are trying far too hard to re-use this method. Instead of passing in command line args for that method to do stuff with it, how about we extract the command line arg handling to something that can be re-used in both places, rather than trying to re-use `handleCommandLineargs`?

File: src/games/strategy/engine/framework/GameRunner.java [hunk 14]
```diff
@@ -666,21 +676,34 @@ public static void joinGame(final GameDescription description, final Messengers
...
-    commands.add("-D" + TRIPLEA_PORT_PROPERTY + "=" + port);
-    commands.add("-D" + TRIPLEA_HOST_PROPERTY + "=" + hostAddressIP);
-    commands.add("-D" + TRIPLEA_NAME_PROPERTY + "=" + messengers.getMessenger().getLocalNode().getName());
-    final String javaClass = "games.strategy.engine.framework.GameRunner";
-    commands.add(javaClass);
-    ProcessRunnerUtil.exec(commands);
+    final boolean startInNewProcess = !("".equals(newClassPath) && MainFrame.getInstance() != null);
+    String prefix = startInNewProcess ? "-D" : "";
+    if (startInNewProcess) {
+      ProcessRunnerUtil.populateBasicJavaArgs(commands, newClassPath);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Note: ProcessRunnerUtil has a lot of problems. It seems like it can be entirely inlined to this class. In the future IMO that would be a really good refactor followed by re-breaking up the class into functional components.

File: src/games/strategy/engine/framework/ProcessRunnerUtil.java (hunk 4 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/ProcessRunnerUtil.java]: The benefit of sending the output, is that I could read debug information like system.out messages and exception stacktraces even after starting those processes.
Abou the blocking: Since we can't interact with it afterwards vie command line, and we know it's triplea we are starting, it's clear that most output will come in lines.
I don't think that something will block there

File: src/games/strategy/engine/framework/GameRunner.java (hunk 13 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/GameRunner.java]: Well actually handle command line args main functionality is to set triplea system properties based on the command line input.
Renaming would be kind of an option, but I don't think that's necessary.

File: src/games/strategy/triplea/ResourceLoader.java (hunk 5 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/ResourceLoader.java]: They must have happened while formatting. Sorry...
Pressing Ctrl+F+Shift makes things easier most of the time.

File: src/games/strategy/triplea/TripleA.java (hunk 7 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/TripleA.java]: You're absolutely right, 
This was needed before I came up with the runnable solution...

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java (hunk 9 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: I don't know if it's really necessary it was like this before. And no I don't think it changes anything

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java (hunk 10 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: Agreed, but I'm trying to simplify this fix, not triplea. This kind of refactoring should be kept for another PR. I will change the necessary stuff though

File: src/games/strategy/triplea/TripleA.java (hunk 6 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/TripleA.java]: I have a better idea... I'll just pass the exception to the startGameMethod...

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java (hunk 11 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: Solved this

**Comment by RoiEXLab** [comment]
@DanVanAtta edited the code to be better. The exception is now properly passed to the original try/catch block (It felt like I did stuff and when I tried the first time it didn't worked, the second time it did... Whatever)
I created an issue to remínd us of the planned refactoring of IGameLoader

**Comment by DanVanAtta** [comment]
Testing is not looking so good. Joining multiple games from lobby now conflicts with each other. Additionally, joining a game, then cancelling, and closing the left over screen via quit closes out the lobby. Before that last item would only close out the one window, and leave the lobby screen open.

File: src/games/strategy/triplea/TripleA.java (hunk 6 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/TripleA.java]: Thinking about this, having a return flag is pretty appropriate. There are a number of cases where a game will not start that are not exceptional. For example, simply not having a map.  A user can trigger that error repeatedly via normal UI controls, therefore it is not an exception since it can be user generated. Passing a runnable is effectively a callback, same thing with exceptions. The return flag is still the same, but the client decides what to do with the response  as part of an if/else block. The control flow is at least explicit.

Failing that, I would consider removing the thing that is causing 'startGame' to swallow exceptions.

File: src/games/strategy/engine/framework/GameRunner.java (hunk 13 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner.java]: Precisely about handle command line args. What I'm suggesting is extracting the key part of handle command line args to a new function, and then have this method and handle command line args both call the common function.

Given testing though, the new process logic might not be quite sound, and a different refactor/simplification could be warranted. Grant it, 'handleCommandLineArgs' is a bit of a rabbit hole for things to clean up..

File: src/games/strategy/engine/framework/startup/mc/ClientModel.java (hunk 9 above)
Comment (RoiEXLab) [inline src/games/strategy/engine/framework/startup/mc/ClientModel.java]: This is not the actual line... deleted the line which call startGameIn new thread from a new thread

**Comment by RoiEXLab** [comment]
@DanVanAtta fixed!

**Comment by RoiEXLab** [comment]
Just in case you are reviewing this too early, I still want to change one or two things...

**Comment by RoiEXLab** [comment]
@DanVanAtta Now, feel free to review!

**Comment by DanVanAtta** [comment]
Thanks for the updates @RoiEXLab . Were you able to test out the lobby / bot interaction of the latest changes? 

I'll have a chance I expect later tonight (about 8 hours from now) to test/review the updates.

**Comment by RoiEXLab** [comment]
This PR prevents creating new processes for every new online game selection ui.
Also users are now disconnected if they haven't downloaded the required map...
Playing multiple Maps at once is still possible.
One thing I noticed, is that closing some Windows like the lobby window shuts down the whole process, but this issue was there even before this PR... We May need to fix that next, but that should be rather easy

**Comment by DanVanAtta** [comment]
> "some Windows like the lobby window shuts down the whole process"

Can you elaborate on this one? It was the case that when playing a multiplayer game, you could close every window down except for the game window, and it would continue. If closing the lobby window closes any games in progress, that sounds new.

**Comment by RoiEXLab** [comment]
Hmm I'll do some testing tomorrow to check if this was there before this PR.
The window was most likely shutting down the process before, but since the game was launched in a different process (which is kind of bad code). It wasn't affected...
@DanVanAtta just in case you have the time I'd like to remind you of the github feature allowing you to edit my branch... 
You could just change the closing behaviour to disposing instead of exiting and we should be good...

**Comment by RoiEXLab** [comment]
I don't want to offend anyone, but the frames logic is sh*t! Everything depends on TripleA starting and stopping multiple processes of itself!

**Comment by RoiEXLab** [comment]
@DanVanAtta Fixed the closing of windows... NEVER EVER AGAIN!
 AFAIK the JavaFX thread stops when all frames are closed. This would have been much easier...

**Comment by DanVanAtta** [comment]
The zombie bot problem, and joining multiple games is fixed, but there is another problem:

- could not join multiple staging screens. Current behavior:
  - join a bot, leave it on the staging screen
  - join a second bot
  - Note: the first staging screen is updated, instead of opening a second staging window


@RoiEXLab I think there needs to be a document on "the state of TripleA" code. It's a project that has evolved over around 12 years. Being an open source project, the development efforts were of varying quality, and particularly it's been a learning project for some at different times. Regardless, there is no disagreement that many things are done 'the hard way'.

**Comment by DanVanAtta** [comment]
For this one. Seems like the fix is to instead of loading over the last staging window, new one up.

yeah, the multi-process launching is really not nice. It has interested effects on the system properties that are used as  crutch to pass information around. (for example the sound muting, it will have a different setting between lobby/staging/game)

**Comment by RoiEXLab** [comment]
> - could not join multiple staging screens. Current behavior:
> 
>     - join a bot, leave it on the staging screen
>     - join a second bot
>     - Note: the first staging screen is updated, instead of opening a second staging window

@DanVanAtta 
I don't quite get what isn't working for you... I can play 2 online games at the same time without problems

**Comment by DanVanAtta** [comment]
You can't for example host two games that are waiting for players, or join multiple games waiting for players.

**Comment by RoiEXLab** [comment]
Ok I'm reverting the only one process feature...
But for the future: I think it would be better to just allow one game per process... If anyone wants to join multiple games at once he should just run triplea twice!

**Comment by RoiEXLab** [comment]
@DanVanAtta You couldn't host 2 games from a single process before (How would you get another window to click on?).
But joining 2 games at the same time is again a thing...

**Comment by DanVanAtta** [comment]
Now I'm confused.

Before we would kick off many new processes (which I agree is pretty terribly). From lobby to staging screen was a new process, and you could have as many staging screens as you want. And then again, after going from staging to in-game, that was yet a new process. 

So I agree it was never the case that you could not host multiple games from the same process, but a key difference is that the staging screen used to also be its own process.

"But joining 2 games at the same time is again a thing..."

So this means you can only be in one game at a time? That sounds even worse than the latest patch I regret to say. Can we revert the original problem that suppressed the exception? Given this ultra-brittle code, it seems that is the best path forward.

**Comment by RoiEXLab** [comment]
@DanVanAtta sorry, I think I explained this wrong...
You _CAN_ join 2 games at the same time again.
No 2 games at the same time still remains possible.
Everthing I believe is now like you want it to be - #1262 is fixed with no side effects and cleaned up code!


The other Part:
I wish that in future versions we only allow 1 game per process.
If a player wants to play multiple games at a time he should just launch a new triplea instance manually, so the triplea codebase is not launching new processes as using them as _advanced threads_

**Comment by RoiEXLab** [comment]
Just in case you are reading your emails.
I edited my previous post to make sense

**Comment by DanVanAtta** [comment]
@RoiEXLab  tested it out, looks like all the use-cases are fixed. I'm going to skip the code review since testing looks good and will spare you the nits.

@RoiEXLab  @ron-murhammer , note: we have a small regression where the error console window pops up again in the case of a missing map. I remember during review cycles I was required a few iterations to clean that up so there would be no error pop-up. Given the use-case, and the previous behavior of 1.8 showing only the error message, we are certainly in an improved state compared to that. Furthermore the more important (regression) problem of not disconnecting after failing to join a game is now fixed.

**Comment by RoiEXLab** [comment]
@DanVanAtta Thanks!
Now I'm going to resolve merge errors on my javafx branch...
